src/arithmetic.jl

```diff
@@ -1,90 +1,60 @@
-multiplicative depth of your algorithm by one and is more expensive to compute. Furthermore,
-one additional rotation is applied with a shift of
-`-sign(shift) * (length(sv) - abs(shift))`.
+Note: If `N` is greater than one, this operation increases the multiplicative level by two,
+otherwise by one.
 
-See also: [`SecureVector`](@ref), [`length`](@ref), [`capacity`](@ref)
-"""
-function Base.circshift(sv::SecureVector, shift::Integer; wrap_by = :capacity)
-    if !(wrap_by in (:length, :capacity))
-        throw(ArgumentError("Unsupported value '$wrap_by' passed to `wrap_by` (must be `:length` or `:capacity`)"))
-    end
+Note: To precompute all required rotation indexes, use `init_array_rotation!`.
 
-    if shift == 0
-        return sv
+See also: [`SecureArray`](@ref), [`init_array_rotation!`](@ref)
+"""
+function Base.circshift(sa::SecureArray{<:AbstractCryptoBackend, N}, shift::NTuple{N, Integer}) where N
+    if all(shift .% size(sa) .== 0)
+        return sa
     end
 
-    rotate(sv, shift; wrap_by)
+    rotate(sa, shift)
 end
 
+"""
+    circshift(sv::SecureVector, shift::Integer)
 
-############################################################################################
-# Matrix
-############################################################################################
-
-# Add
-Base.:+(sm1::SecureMatrix, sm2::SecureMatrix) = add(sm1, sm2)
-Base.:+(sm::SecureMatrix, pm::PlainMatrix) = add(sm, pm)
-Base.:+(pm::PlainMatrix, sm::SecureMatrix) = add(sm, pm)
-Base.:+(sm::SecureMatrix, scalar::Real) = add(sm, scalar)
-Base.:+(scalar::Real, sm::SecureMatrix) = add(sm, scalar)
-
-# Subtract
-Base.:-(sm1::SecureMatrix, sm2::SecureMatrix) = subtract(sm1, sm2)
-Base.:-(sm::SecureMatrix, pm::PlainMatrix) = subtract(sm, pm)
-Base.:-(pm::PlainMatrix, sm::SecureMatrix) = subtract(pm, sm)
-Base.:-(sm::SecureMatrix, scalar::Real) = subtract(sm, scalar)
-Base.:-(scalar::Real, sm::SecureMatrix) = subtract(scalar, sm)
-
-# Negate
-Base.:-(sm::SecureMatrix) = negate(sm)
+Circularly shift, i.e., rotate the data in `sv` by `shift` positions, similarly to Julia's
+`circshift` for regular arrays.
 
-# Multiply
-Base.:*(sm1::SecureMatrix, sm2::SecureMatrix) = multiply(sm1, sm2)
-Base.:*(sm::SecureMatrix, pm::PlainMatrix) = multiply(sm, pm)
-Base.:*(pm::PlainMatrix, sm::SecureMatrix) = multiply(sm, pm)
-Base.:*(sm::SecureMatrix, scalar::Real) = multiply(sm, scalar)
-Base.:*(scalar::Real, sm::SecureMatrix) = multiply(sm, scalar)
+Note: This operation increases the multiplicative level by one.
 
-# Circular shift
-function Base.circshift(sm::SecureMatrix, shift::Tuple{Integer, Integer})
-    if shift[1] % size(sm, 1) == 0 && shift[2] % size(sm, 2) == 0
-        return sm
-    end
+Note: To precompute all required rotation indexes, use `init_array_rotation!`.
 
-    rotate(sm, shift)
+See also: [`SecureVector`](@ref), [`init_array_rotation!`](@ref)
+"""
+function Base.circshift(sv::SecureVector, shift::Integer)
```

Comment:
Why a second implementation for SV? Shouldn't this be covered by the more general one above?